src/DataTypes/Serializations/SerializationVariant.cpp

```diff
@@ -465,8 +482,36 @@ void SerializationVariant::deserializeBinaryBulkWithMultipleStreams(
 
     settings.path.pop_back();
 
-    /// Second, calculate limits for each variant by iterating through new discriminators
+    /// Second, calculate offsets and limits for each variant by iterating through new discriminators
     /// if we didn't do it during discriminators deserialization.
+    if (variant_rows_offsets.empty())
+    {
+        variant_rows_offsets.resize(variants.size(), 0);
+
+        if (rows_offset)
+        {
+            size_t discriminators_offset = col.getLocalDiscriminatorsPtr()->size() - limit - rows_offset;
```

Comment:
We don't guarantee that we will read exactly `limit` rows in discriminators, we can read less. See comments:
https://github.com/ClickHouse/ClickHouse/blob/7ba7dd4323267769bac30c69cec837e21e5278b0/src/DataTypes/Serializations/SerializationVariant.cpp#L475-L479
https://github.com/ClickHouse/ClickHouse/blob/7ba7dd4323267769bac30c69cec837e21e5278b0/src/DataTypes/Serializations/SerializationVariantElement.cpp#L146-L147

I don't remember the case when it can be less to be honest, but let's assume it can happen just in case

Reply:
@Avogar, how we determine the actual number of rows read when offsets are in the cache. In SerializationVariant.cpp, the code currently uses the limit directly as the number of rows to read, however we can actually read less than limit discriminators.

Reply:
I think the best way is to save the size of the column at the end of deserialization in the deserialization state. In this case in next deserialization we can compare the new size and the previous size and get the number of new discriminators. We can set it to 0 when at the start of deserialization the size of the column is 0.